tests: syscall: add support for string_alloc_copy()

LixinGuoX opened the pull request:
Add support for validating string_alloc_copy() API with invalid
string parameter.

Signed-off-by: Lixin Guo <lixinx.guo@intel.com>

LixinGuoX commented:
Maybe some bug issue with several boards.

LixinGuoX commented:
CI got blocked for #40870, #40873, #40874

enjiamai commented on tests/kernel/mem_protect/syscalls/src/main.c:
```diff
@@ -253,6 +254,9 @@ void test_user_string_alloc_copy(void)
 	zassert_equal(ret, 0, "string should have matched");
+
+	ret = string_alloc_copy(src);
+	zassert_equal(ret, -1, "got %d", ret);
```
could the -1 be instead of -EPERM here?

LixinGuoX replied:
the return value is defined in z_vrfy_string_alloc_copy() and z_impl_string_alloc_copy(), so I think it is Inappropriate to use `-EPERM` here.
```
int z_impl_string_alloc_copy(char *src)
{
	if (!strcmp(src, kernel_string)) {
		return 0;
	} else {
		return -2;
	}
}

static inline int z_vrfy_string_alloc_copy(char *src)
{
	char *src_copy;
	int ret;

	src_copy = z_user_string_alloc_copy((char *)src, BUF_SIZE);
	if (!src_copy) {
		return -1;
	}

	ret = z_impl_string_alloc_copy(src_copy);
	k_free(src_copy);

	return ret;
}
```

enjiamai replied:
understand, thx!

enjiamai commented on tests/kernel/mem_protect/syscalls/testcase.yaml:
```diff
@@ -2,3 +2,4 @@ tests:
   kernel.memory_protection.syscalls:
     filter: CONFIG_ARCH_HAS_USERSPACE
     tags: kernel security userspace ignore_faults
+    platform_exclude: qemu_cortex_r5 mps2_an521_ns
```
Could you tell me why we skip these two platforms?

LixinGuoX replied:
These two boards has bug issue with processing null parameter, so exclude them temporarily.
see #40873, #40874

theotherjimmy replied:
This should probably be done with a better filter. Since this test needs null pointer exception detection, we should filter that out here instead of relying on platform excludes.

Commit:
tests: syscall: add support for string_alloc_copy()

Add support for validating string_alloc_copy() API with invalid
string parameter.

Signed-off-by: Guo Lixin <lixinx.guo@intel.com>

LixinGuoX commented:
> I recommend adding the following patch:
> 
> ```diff
> From 24a09f13f63c90a9049661cb64e775695eb1d512 Mon Sep 17 00:00:00 2001
> From: Jimmy Brisson <jimmy.brisson@linaro.org>
> Date: Mon, 24 Jan 2022 12:00:47 -0600
> Subject: [PATCH] Skip platforms that do not support null pointer checks
> 
> Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>
> ---
>  tests/kernel/mem_protect/syscalls/testcase.yaml | 2 +-
>  1 file changed, 1 insertion(+), 1 deletion(-)
> 
> diff --git a/tests/kernel/mem_protect/syscalls/testcase.yaml b/tests/kernel/mem_protect/syscalls/testcase.yaml
> index 6c40fcda00..ab8c03ed4f 100644
> --- a/tests/kernel/mem_protect/syscalls/testcase.yaml
> +++ b/tests/kernel/mem_protect/syscalls/testcase.yaml
> @@ -1,4 +1,4 @@
>  tests:
>    kernel.memory_protection.syscalls:
> -    filter: CONFIG_ARCH_HAS_USERSPACE
> +    filter: CONFIG_ARCH_HAS_USERSPACE and not CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE
>      tags: kernel security userspace ignore_faults
> -- 
> 2.33.1
> ```
> 
> Notes from debugging this are here: https://theotherjimmy.github.io/blog/blog/2022-01-24-bug-notes-str-alloc-copy.html

Thank you for your information about the issue on this, this is a relatively perfect solution, I have updated as recommended, thx!

LixinGuoX commented:
Lone time no update, so close it.
